md/groovy-in-bonita.md

```diff
@@ -0,0 +1,347 @@
+
+`*` for mandatory fields
+
+```groovy
+Comment {
+    String content*
+    AppUser createdBy* //Aggregation
+    DateTime creationDate*
+    AppUser lastEditedBy* //Aggregation
+}
+
+AppUser {
+    String firstName*
+    String lastName*
+    Address address //Composition
+}
+
+Address {
+    Street street* //Composition
+    String postalCode*
+    String city*
+}
+ 
+Street {
+    String name*
+    String number*
+    String additionalInfo
+}
+```
+
+### Instantiate a Business Data
+
+#### Using the generated DAO
+
+By default a `newInstance` factory method is generated in the object DAO.  
+This method has as many parameters as mandatory fields for this object.
+
+```groovy
+appUserDAO.newInstance('Jane','Doe') // create a UserApp instance with firstName = 'Jane' and lastName = 'Doe'
+```
+
+#### Using a constructor with named arguments
+
+```groovy
+new Address(street: new Street(number: '32', name: 'Gustave Eiffel'), postalCode: '38000', city: 'Grenoble')
+```
+
+#### From a map
+
+Given a complex contract input
+
+```
+userInput COMPLEX
+	firstName TEXT
+	lastName TEXT
+```
+
+As a COMPLEX input is stored as Map it is possible to write
+
+```groovy
+userInput as UserApp
+```
+
+### List Business Objects using DAO
+
+When defining your BDM you can write custom queries in JPQL than can be called using the object DAO.
+Some queries are generated by default like: `find`, `findByFirstName`, `findByLastName`...
+
+All object DAO are injected at runtime in Groovy script expression or can be retrieved using [APIClient#getDAO](https://documentation.bonitasoft.com/javadoc/api/${varVersion}/org/bonitasoft/engine/api/APIClient.html#getDAO-java.lang.Class-)
+
+Here is the UserAppDAO interface generated for the UserApp object:
+
+```java
+public interface UserAppDAO extends BusinessObjectDAO {
+    public UserApp findByPersistenceId(Long persistenceId);
+
+    public List<UserApp> findByFirstName(String firstName, int startIndex, int maxResults);
+
+    public List<UserApp> findByLastName(String lastName, int startIndex, int maxResults);
+
+    public List<UserApp> find(int startIndex, int maxResults);
+
+    public Long countForFindByFirstName(String firstName);
+
+    public Long countForFindByLastName(String lastName);
+
+    public Long countForFind();
+
+    public UserApp newInstance(String firstName, String lastName);
+}
+```
+
+So, in a Groovy script, you can access the data like this:  
+
+```groovy
+def users = userAppDAO.find(0, 10) // returns to first tens users ordered by persistenceId
+def johnUsers = userAppDAO.findByFirstName('John', 0, 10) // returns to first tens users with firstName == 'John' ordered by persistenceId
+```
+
+### Update a Business Data
+
+The example below is the generated code when editing a UserApp address from a contract
+
+`user` is the existing data in the process  
+`userInput` is the contract input of the edition task
+
+```groovy
+if (!userInput?.address) { // As Address is not mandatory it can be null
```

Comment:
Maybe a beginner needs to know that in Groovy, Object boolean evaluation is different that in java: 
_A null object reference evaluates to false, a non null reference to true, but for for strings, collections, iterators and enumerations it also takes into account the size._
-> It is the first boolean evaluation in this tutorial, but maybe a line about this important Groovy behavior in the general section would be nice?

Reply:
yes, I'll add section about Groovy Thruth